stl/inc/ranges

```diff
@@ -636,12 +636,138 @@ namespace ranges {
+        // clang-format on
+    public:
+        constexpr _Non_propagating_cache() noexcept : _Dummy{} {}
+
+        _CONSTEXPR20 ~_Non_propagating_cache() noexcept {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Construct_in_place(_Dummy);
+            }
+        }
+
+        // clang-format off
+        ~_Non_propagating_cache() requires is_trivially_destructible_v<_Ty> = default;
+        // clang-format on
+
+        constexpr _Non_propagating_cache(const _Non_propagating_cache&) noexcept : _Dummy{} {}
+
+        _CONSTEXPR20 _Non_propagating_cache(_Non_propagating_cache&& _Other) noexcept : _Dummy{} {
+            if (_Other._Engaged) {
+                _Other._Val.~_Ty();
+                _Construct_in_place(_Other._Dummy);
+                _Other._Engaged = false;
+            }
+        }
+
+        _CONSTEXPR20 _Non_propagating_cache& operator=(const _Non_propagating_cache& _Other) noexcept {
+            if (_STD addressof(_Other) == this) {
+                return *this;
+            }
+
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Construct_in_place(_Dummy);
+                _Engaged = false;
+            }
+
+            return *this;
+        }
+
+        _CONSTEXPR20 _Non_propagating_cache& operator=(_Non_propagating_cache&& _Other) noexcept {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Construct_in_place(_Dummy);
+                _Engaged = false;
+            }
+
+            if (_Other._Engaged) {
+                _Other._Val.~_Ty();
+                _Construct_in_place(_Other._Dummy);
+                _Other._Engaged = false;
+            }
+
+            return *this;
+        }
+
+        _NODISCARD constexpr _Ty& operator*() noexcept {
+            _STL_INTERNAL_CHECK(_Engaged);
+            return _Val;
+        }
+        _NODISCARD constexpr const _Ty& operator*() const noexcept {
+            _STL_INTERNAL_CHECK(_Engaged);
+            return _Val;
+        }
+
+        template <class _It>
+        _CONSTEXPR20 _Ty& _Emplace_deref(_It& _Iter) noexcept(noexcept(_Ty(*_Iter))) /* strengthened */ {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Construct_in_place(_Dummy);
+                _Engaged = false;
+            }
+
+            _Construct_in_place(_Val, *_Iter);
+            _Engaged = true;
+
+            return _Val;
+        }
+
+    private:
+        union {
+            _Nontrivial_dummy_type _Dummy;
+            _Ty _Val;
+        };
+        bool _Engaged = false;
+    };
+
+    // clang-format off
+    template <_Destructible_object _Ty>
+        requires is_trivially_destructible_v<_Ty> && is_default_constructible_v<_Ty>
+    class _Non_propagating_cache<_Ty> {
+        // provide the same API more efficiently for trivially destructible and default constructible types
         // clang-format on

```

Comment:
I don't understand why this needs default construction - and in any event I don't think anything in the wording permits calling a default constructor that can have arbitrary side effects.

Reply:
Burninated and left for later